File: AOC_18/Day4/Day4.py
import datetime


class Action:
    def __init__(self, action_string):
        temp = action_string.split(']')
        time = temp[0].split(' ')
        date = time[0].split('-')
        time = time[1].split(':')
        self.date = datetime.datetime(int(date[0][1:]), int(date[1]), int(date[2]), int(time[0]), int(time[1]))
        if temp[1][1] == 'G':
            self.action = int(temp[1].split(' ')[2][1:])
        elif temp[1][1] == 'f':
            self.action = -1
        else:
            self.action = -2


actions = []
guard_data = {}
for line in open('input').readlines():
    action = Action(line)
    if action.action >= 0 and action.action not in guard_data.keys():
        guard_data[action.action] = [0]*60
    actions.append(action)


curr_guard = -1
asleep = False
minute = 0
for action in sorted(actions, key=lambda x: x.date):
    if asleep:
        for i in range(minute, action.date.minute):
            guard_data[curr_guard][i] += 1
    if action.action == -2:
        asleep = False
    elif action.action == -1:
        asleep = True
    else:
        curr_guard = action.action
    minute = action.date.minute

curr_guard = max(guard_data, key=lambda x: sum(guard_data[x]))
print("Part 1:", guard_data[curr_guard].index(max(guard_data[curr_guard])) * curr_guard)

curr_guard = max(guard_data, key=lambda x: max(guard_data[x]))
print("Part 2:", guard_data[curr_guard].index(max(guard_data[curr_guard]))*curr_guard)

File: AOC_18/Day4/Day4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strconv"
	"time"
)

type Action int

const (
	Wake Action = iota
	Asleep
	Shift
)

type Obsv struct {
	action Action
	guard  int
	date   time.Time
}

func main() {
	file, err := os.Open("Day4/input")
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		os.Exit(1)
	}
	scanner := bufio.NewScanner(file)

	observations := []Obsv{}
	guard_data := make(map[int][]int)

	for scanner.Scan() {
		obsv := scanner.Text()
		date, _ := time.Parse("2006-01-02 15:04", obsv[1:17])

		var action Action
		switch obsv[19] {
		case 'G':
			action = Shift
		case 'w':
			action = Wake
		case 'f':
			action = Asleep
		}

		obsvStruct := Obsv{date: date, action: action}
		if action == Shift {
			r := regexp.MustCompile(`\d+`)
			obsvStruct.guard, _ = strconv.Atoi(r.FindString(obsv[26:]))
			if _, present := guard_data[obsvStruct.guard]; !present {
				guard_data[obsvStruct.guard] = make([]int, 60)
			}
		}

		observations = append(observations, obsvStruct)
	}

	sort.Slice(observations, func(i, j int) bool {
		return observations[i].date.Before(observations[j].date)
	})

	curr_guard := observations[0].guard
	asleep := false
	minute := 0
	for _, obsv := range observations {
		if asleep {
			for i := minute; i < obsv.date.Minute(); i++ {
				guard_data[curr_guard][i]++
			}
		}
		if obsv.action == Asleep {
			asleep = true
		} else if obsv.action == Wake {
			asleep = false
		} else {
			curr_guard = obsv.guard
		}
		minute = obsv.date.Minute()
	}

	most_sleep_total_guard := -1
	most_sleep_total_val := 0
	most_sleep_minute_guard := -1
	most_sleep_minute := 0
	most_sleep_minute_val := 0
	for guard, minutes := range guard_data {
		sum := 0
		for i, min := range minutes {
			sum += min
			if min > most_sleep_minute_val {
				most_sleep_minute_val = min
				most_sleep_minute = i
				most_sleep_minute_guard = guard
			}
		}
		if sum > most_sleep_total_val {
			most_sleep_total_val = sum
			most_sleep_total_guard = guard
		}
	}

	max := 0
	for i, min := range guard_data[most_sleep_total_guard] {
		if min > guard_data[most_sleep_total_guard][max] {
			max = i
		}
	}
	fmt.Println("Part 1:", max*most_sleep_total_guard)
	fmt.Println("Part 2:", most_sleep_minute*most_sleep_minute_guard)
}
